File: apps/python-server/src/cron/analyze_call.py
import io
import requests
import soundfile as sf
from ..utils.vapi_client import create_vapi_client
from ..utils.openai_client import create_openai_client
from ..config import settings
from ..utils.create_mono_buffer import create_mono_buffer
from ..utils.create_segments import create_segments_from_words

async def analyze_call(call_id: str):
    """Analyze a call recording and generate transcripts."""
    # Initialize clients
    vapi_client = create_vapi_client(settings.vapi_api_key)
    openai_client = create_openai_client(settings.openai_api_key)
    
    # Get call details from Vapi
    call = vapi_client.calls.get(call_id)
    
    if not call.artifact or not call.artifact.stereo_recording_url:
        return

    # Download the audio file
    response = requests.get(call.artifact.stereo_recording_url)
    audio_data = io.BytesIO(response.content)
    
    # Load audio file using soundfile
    data, sample_rate = sf.read(audio_data)
    
    # Split channels
    left_channel = data[:, 0]
    right_channel = data[:, 1]
    
    # Create and save mono buffers for each channel
    left_buffer = create_mono_buffer(left_channel, sample_rate)
    right_buffer = create_mono_buffer(right_channel, sample_rate)
    
    # Save files locally for testing
    with open(f"temp_{call_id}_left.wav", "wb") as f:
        f.write(left_buffer.getvalue())
    with open(f"temp_{call_id}_right.wav", "wb") as f:
        f.write(right_buffer.getvalue())
        
    # Reset buffer positions for subsequent use
    left_buffer.seek(0)
    right_buffer.seek(0)
    
    # Transcribe left channel
    left_transcript = openai_client.audio.transcriptions.create(
        file=("audio.wav", left_buffer, "audio/wav"),
        model="whisper-1",
        language="en",
        response_format="verbose_json",
        timestamp_granularities=["word"],
    )
    
    # Transcribe right channel
    right_transcript = openai_client.audio.transcriptions.create(
        file=("audio.wav", right_buffer, "audio/wav"),
        model="whisper-1",
        language="en",
        response_format="verbose_json",
        timestamp_granularities=["word"],
    )
    
    original_transcript = call.artifact.transcript

    # In your analyze_call function:
    segments = create_segments_from_words(left_transcript.words, right_transcript.words)
    print("\n".join([segment["text"] for segment in segments]))

    print("\n=== Original Transcript ===")
    print(original_transcript)

File: apps/socket-server/src/cron/analyzeCall.ts
import vapi from "../utils/vapiClient";
import openai from "../utils/openAiClient";
import { AudioContext } from "web-audio-api";
// ... existing imports ...

export const analyzeCall = async (callId: string) => {
  const call = await vapi.calls.get(callId);
  if (!call.artifact?.stereoRecordingUrl) {
    return;
  }

  const response = await fetch(call.artifact?.stereoRecordingUrl);
  const arrayBuffer = await response.arrayBuffer();
  const audioContext = new AudioContext();
  const audioBuffer = await audioContext.decodeAudioData(arrayBuffer);

  // Get individual channels
  const leftChannel = audioBuffer.getChannelData(0);
  const rightChannel = audioBuffer.getChannelData(1);

  // If you need to create separate audio files from these channels
  const createMonoBuffer = (channel: Float32Array) => {
    const monoBuffer = audioContext.createBuffer(
      1,
      channel.length,
      audioBuffer.sampleRate,
    );
    monoBuffer.getChannelData(0).set(channel);
    return monoBuffer;
  };

  const leftBuffer = createMonoBuffer(leftChannel);
  const rightBuffer = createMonoBuffer(rightChannel);

  // Convert AudioBuffer to Blob
  const audioBlob = await audioBufferToBlob(leftBuffer);
  const audioFile = new File([audioBlob], "audio.wav", { type: "audio/wav" });

  const leftTranscript = await openai.audio.transcriptions.create({
    file: audioFile,
    model: "whisper-1",
    language: "en",
    response_format: "verbose_json",
    timestamp_granularities: ["word"],
  });

  const audioBlobRight = await audioBufferToBlob(rightBuffer);
  const audioFileRight = new File([audioBlobRight], "audio.wav", {
    type: "audio/wav",
  });

  const rightTranscript = await openai.audio.transcriptions.create({
    file: audioFileRight,
    model: "whisper-1",
    language: "en",
    response_format: "verbose_json",
    timestamp_granularities: ["word"],
  });

  const originalTranscript = call.artifact?.transcript;

  console.log("originalTranscript", originalTranscript);
  console.log("openaitranscript", leftTranscript);
  console.log("openaitranscript", rightTranscript);
};

async function audioBufferToBlob(audioBuffer: AudioBuffer): Promise<Blob> {
  const wav = await encodeWAV(audioBuffer);
  return new Blob([wav], { type: "audio/wav" });
}

function encodeWAV(audioBuffer: AudioBuffer): Buffer {
  const numChannels = audioBuffer.numberOfChannels;
  const sampleRate = audioBuffer.sampleRate;
  const format = 1; // PCM
  const bitDepth = 16;
  const data = audioBuffer.getChannelData(0);

  const buffer = Buffer.alloc(44 + data.length * 2);

  // WAV header
  buffer.write("RIFF", 0);
  buffer.writeUInt32LE(36 + data.length * 2, 4);
  buffer.write("WAVE", 8);
  buffer.write("fmt ", 12);
  buffer.writeUInt32LE(16, 16);
  buffer.writeUInt16LE(format, 20);
  buffer.writeUInt16LE(numChannels, 22);
  buffer.writeUInt32LE(sampleRate, 24);
  buffer.writeUInt32LE((sampleRate * numChannels * bitDepth) / 8, 28);
  buffer.writeUInt16LE((numChannels * bitDepth) / 8, 32);
  buffer.writeUInt16LE(bitDepth, 34);
  buffer.write("data", 36);
  buffer.writeUInt32LE(data.length * 2, 40);

  // Audio data
  let offset = 44;
  for (let i = 0; i < data.length; i++) {
    const sample = Math.max(-1, Math.min(1, data[i]));
    buffer.writeInt16LE(sample * 0x7fff, offset);
    offset += 2;
  }

  return buffer;
}
